File: mergesort.ml
let divide l =
  let rec _divide l acc mid = match l with
      [] -> (List.rev acc, l)
    | l when mid <= 0 -> (List.rev acc, l)
    | e::l -> _divide l (e::acc) (mid-1)
  in
  _divide l [] ((List.length l) / 2) ;;

let rec merge l1 l2 = match (l1, l2) with
    (l, []) | ([], l) -> l
  | (e1::l1, e2::l2) ->
      if e1 > e2
      then e2 :: (merge (e1::l1) l2)
      else e1 :: (merge l1 (e2::l2)) ;;

let rec mergesort = function
    [] -> []
  | [e1] -> [e1]
  | [e1;e2] -> if e1 > e2 then [e2;e1] else [e1;e2]
  | l -> let l1, l2 = divide l in merge (mergesort l1) (mergesort l2) ;;

List.iter (fun e -> print_int e; print_string " ") (mergesort [5;9;1;3;4;6;6;3;2]) ;
print_newline () ;;

File: mergesort.cpp

#include <iostream>
#include <vector>
using namespace std;

void merge(int a[], int start, int mid, int end){
    vector <int> b1, b2;
    for(int i = start; i <= mid; i++){
        b1.push_back(a[i]);
    }
    for(int i = mid + 1; i <= end; i++){
        b2.push_back(a[i]);
    }
    int p1 = 0, p2 = 0, p3 = start;
    while(p1 < b1.size() && p2 <b2.size()){
        if(b1[p1] < b2[p2]){
            a[p3++] = b1[p1++];
        } else {
            a[p3++] = b2[p2++];
        }
    }
    while(p1 < b1.size()){
        a[p3++] = b1[p1++];
    }
    while(p2 < b2.size()){
        a[p3++] = b2[p2++];
    }
}

void mergesort(int a[], int start, int end){
    if (start < end){
        int mid = (start + end)/2;
        mergesort(a, start, mid);
        mergesort(a, mid + 1, end);
        merge(a, start, mid, end);
    }
}


signed main() {
    int a[] = {5,9,1,3,4,6,6,3,2};
    mergesort(a, 0, sizeof(a)/sizeof(int) - 1);
    for (int f: a){
        cout << f << " ";
    }
    cout << endl; 
}
